### PR TITLE
implemented turretControls and propulsion

### PR DESCRIPTION
**Pull request by wu-benjamin**
- Added ability to use space bar to tap shoot torpedos with manual ship control
- Implemented `turretControls` functions `aimTurret`, `getNumberOfTubes`, `getTubeCooldown`, and `fireTorpedo` with proper response format
- Added separate cooldowns for ship torpedo tubes
- Implemented optional duration (in frames) for torpedos to explode after the duration if no collision happens; default to infinite fuse if default value of 0 is passed into torpedo constructor
- Rotated torpedo image so the angle of travel also sets the orientation of the sprite correctly
- Added explosion animation for torpedo (hit-box still exists during this) with configurable explosion duration and size (the explosion balloons with the square of time elapsed; not completely sure if that is what actually happens in an explosion but it seems/looks plausible to me)
- Added physics for thrusters and `setThruster` with proper response format

### CODE REVIEW

File: src/ship/turretControls.js [hunk 1]
```diff
@@ -1,20 +1,71 @@
...
-		// check for valid torpedo stuff, then create new one
-		// torpedo = new Torpedo(new Vector2(40, 30), this.pos, this.game);
-		// this.game.delObjects.push(torpedo);
+		//check for valid torpedo stuff, then create new one
+		if (tubeIndex < 0 || tubeIndex >= NUMBER_OF_TUBES) {
+			let errorMessage = "fireTorpedo due to invalid tube index; expected tubeIndex from 0 (inclusive) up to " + NUMBER_OF_TUBES + " (exclusive) but received " + tubeIndex;
+			console.log(errorMessage);
+			return new response(400, [errorMessage], {}, false);
+		}
+		let tubeCooldownResponse = this.getTubeCooldown(tubeIndex);
```
Comment (bjonli) [inline src/ship/turretControls.js:55]: const?

File: src/ship/torpedo.js [hunk 2]
```diff
@@ -1,40 +1,57 @@
...
+		if (this.fuseFrameDuration > 0 && this.parentShip.game.frame - this.frameCreated >= this.fuseFrameDuration) {
+			this.explode();  // Explode after fuse expires
+		}
+		if (this.hasExploded) {
+			let currentFrame = this.parentShip.game.frame;
+			let framesSinceExploded = currentFrame - this.frameExploded;
+			let explosionNumber = Math.floor(framesSinceExploded * NUMBER_OF_EXPLOSION_SPRITES / FRAMES_FOR_EXPLOSION);
+			if (explosionNumber < NUMBER_OF_EXPLOSION_SPRITES) {
+				this.image = this.game.images["explosion" + explosionNumber];
+				this.size = this.originalSize.scale(EXPLOSION_MAX_SCALE * Math.sqrt(framesSinceExploded / FRAMES_FOR_EXPLOSION));
```
Comment (bjonli) [inline src/ship/torpedo.js:41]: nit pick, can we store the radius changes in some sort of static map; not a fan of how we recalculate everything.

File: src/ship/turretControls.js [hunk 3]
```diff
@@ -1,20 +1,71 @@
...
 	}
+	// This is currently assumed to be an absolute direction, it can be implemented as a relative direction through a change of basis
 	aimTurret(aimTo){
 		//User called function for aiming turret
+		let newDirection = aimTo.normalize();
+		if (newDirection.magnitude() > 0) {
+			this.direction = newDirection;
+			return new response(200, [], {}, true);
+		} else {
+			let errorMessage = "aimTurret failed due to zero aimTo and undefined direction; no update to aim made";
```
Comment (bjonli) [inline src/ship/turretControls.js:27]: cosnt

File: src/ship/turretControls.js [hunk 4]
```diff
@@ -1,20 +1,71 @@
...
+			console.log(errorMessage);
+			return new response(400, [errorMessage], {}, false);
+		}
+	}
+	getNumberOfTubes(){
+		return new response(200, [], {numberOfTubes: this.numberOfTubes}, true);
 	}
 	getTubeCooldown(tubeIndex){
+		if (tubeIndex < 0 || tubeIndex >= NUMBER_OF_TUBES) {
+			let errorMessage = "getTubeCooldownFailed due to invalid tube index; expected tubeIndex from 0 (inclusive) up to " + NUMBER_OF_TUBES + " (exclusive) but received " + tubeIndex;
```
Comment (bjonli) [inline src/ship/turretControls.js:37]: const

**Review by bjonli — COMMENTED** [review]
small things, but otherwise lgtm

File: src/ship/turretControls.js (hunk 1 above)
Comment (wu-benjamin) [reply src/ship/turretControls.js]: updated

File: src/ship/torpedo.js (hunk 2 above)
Comment (wu-benjamin) [reply src/ship/torpedo.js:41]: There would have to be an entry per frame in the explosion duration—lets leave that change for later

File: src/ship/turretControls.js (hunk 4 above)
Comment (wu-benjamin) [reply src/ship/turretControls.js]: updated

File: src/ship/turretControls.js (hunk 3 above)
Comment (wu-benjamin) [reply src/ship/turretControls.js]: updated

**Comment by zrwaite** [comment]
I am probably going to do controller testing and reconfiguration with Josiah today, so I might change a lot of the manual stuff there, but looks all good!

**Pull request merged**